app/controllers/concerns/foreman/controller/registration.rb

```diff
@@ -27,7 +28,11 @@ def global_registration_vars
                    organization: (organization || User.current.default_organization || User.current.my_organizations.first),
                    location: (location || User.current.default_location || User.current.my_locations.first),
                    hostgroup: host_group,
-                   operatingsystem: operatingsystem })
+                   operatingsystem: operatingsystem,
+                   url_host: url.host,
+                   registration_url: register_url(host: url.host, protocol: url.scheme),
+                   registration_url_with_uuid: register_url(host: url.host, protocol: url.scheme, uuid: 'UUID').sub('UUID', '$UUID'),
```

Comment:
We still have support for `RAILS_RELATIVE_URL_ROOT`. If that's used, the Proxy URL would break.

Also note that you don't take the port into consideration so you would lose that info.

Please write some test for the use case where a param is passed in.

Reply:
Do we really support that? I didn't find a mention in our manual or an installer options. Looking at the puppet-foreman, it seems to be based on `foreman_url` parameter, however it's documentation does not mention such behavior.

Anyway, the `register_url` respects that variable, link any other helper. E.g.

```
RAILS_RELATIVE_URL_ROOT=/foreman be rails c
[1] pry(main)> app.register_path
=> "/foreman/register"
[2] pry(main)> app.register_url
=> "http://www.example.com/foreman/register"
[3] pry(main)> app.register_url host: 'example.tst'
=> "http://example.tst/foreman/register"
```

I agree we need to add port as well though. Test would be great for sure.

Reply:
Ok, finally found undocummented parameter you can use

```
register_url host: 'example.tst', script_name: ''
=> "http://example.tst/register"
```

so I think required changes are:
1) use empty script_name in case of the url is coming from the parameter *sigh*
2) also pass port parsed out of the url
3) add a test for generating the right values to the variables in all these cases

Thanks for bearing with us.

Reply:
> Do we really support that? I didn't find a mention in our manual or an installer options. Looking at the puppet-foreman, it seems to be based on `foreman_url` parameter, however it's documentation does not mention such behavior.

It's defined here:
https://github.com/theforeman/foreman/blob/93dd5ad7657023873c87c7aabeba4e9ed23f713e/config.ru#L9

Then it can be set here by the installer:
https://github.com/theforeman/puppet-foreman/blob/495727bdfe4a60df28c542cf5b771a91e20433b3/templates/_suburi.conf.erb#L1

I can't say for sure it still works with Puma out of the box.

However, the UI has repeatedly broken this and I haven't seen anyone complain. Perhaps we should open an RFC to remove it?

> The problem is, in such configratuion, the prefix _will be added_ even for proxy url, I didn't find a way to ignore such prefix yet

That was precisely my point that I should have been explicit about: the Foreman proxy URL should not be based on the Rails URL helper because they're not really related. I still am strongly in favor of using proper URL helpers rather than hardcoding, but with a remote API that's hard to avoid.

I think here it should be:
```ruby
if params[:url].presence
  uri = URI(params[:url])
  uri.path += '/register'
  registration_url = uri.to_s
  uri.query = 'uuid=$UUID'
  registration_url_with_uuid = uri.to_s
else
  registration_url = register_url
  registration_url_with_uuid = register_url(uuid: 'UUID').sub('UUID', '$UUID')
end
```

Reply:
How about `Setting[:foreman_url]`, isn't that more important than ENV['RAILS_RELATIVE_URL_ROOT']?
What if they don't have same value, what should be expected behavior?

Reply:
I know it's in the code. As I said, the installer only does that if it sees the suburi in [forman-url](https://github.com/theforeman/puppet-foreman/blob/6f5b99bb08f19ff61410dc6208346b9c8c333221/manifests/config/apache.pp#L137-L152)  however foreman-url documentation [does not mention](https://github.com/theforeman/puppet-foreman/blob/6f5b99bb08f19ff61410dc6208346b9c8c333221/manifests/init.pp#L43) that at all. 

> However, the UI has repeatedly broken this and I haven't seen anyone complain. Perhaps we should open an RFC to remove it?

Yes! Based on today status, I don't think people know about that and I'd also say it's possible but not supported.

> That was precisely my point that I should have been explicit about: the Foreman proxy URL should not be based on the Rails URL helper because they're not really related. I still am strongly in favor of using proper URL helpers rather than hardcoding, but with a remote API that's hard to avoid.

Yeah, got it a second after I sent the first post, then I editted it. I think you are right with your example, however the query part is not that easy. While it would avoid the escaping issue, it overrides any query already present. We'd need to split it by `&`, add a new to the array and the join again and set it only then. In other words, it has the same flaw as just appending, we only get rid of the `?` for the price of more complex code and two different paths handling it. I don't think it's worth it.

Reply:
> How about Setting[:foreman_url], isn't that more important than ENV['RAILS_RELATIVE_URL_ROOT']?
What if they don't have same value, what should be expected behavior?

The installer I believe takes care of their values. If user changes that manually, that's a problem. I think in this case, we should either use `Setting[:foreman_url]` as a primary source. If there's external URL, we should ignore the ENV variable, since that's only for internal URLs.